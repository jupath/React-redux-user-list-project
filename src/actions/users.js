
export const usersAreLoading = bool => ({
  type: 'USERS_ARE_LOADING',
  isLoading: bool,
});

export const setUsers = users => ({
  type: 'SET_USERS',
  users,
});

export const startSetUsers = url => async (dispatch) => {
  dispatch(usersAreLoading(true));

  try {
    const response = await fetch(url);
    const data = await response.json();
    const results = await data.results;
    const users = await results.map(user => ({
      username: user.login.username,
      firstName: user.name.first,
      lastName: user.name.last,
      email: user.email,
      address: `${user.location.street}, ${user.location.city}, ${user.location.state}, ${user.location.postcode}`,
      dob: user.dob,
      cell: user.cell,
      image: user.picture.large,
    }));
    await dispatch(usersAreLoading(false));
    await dispatch(setUsers(users));
  } catch (e) {
    console.error(e);
  }
};

export const updateUser = (username, updatedData) => ({
  type: 'UPDATE_USER',
  username,
  updatedData,
});
