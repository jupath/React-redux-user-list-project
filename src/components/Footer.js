import React from 'react';

const Footer = () => (
  <footer className="footer pb-4 text-center">
    <h5>View <a href="https://github.com/jupath/React-redux-user-list-project/" target="_blank" rel="noopener noreferrer" className="font-weight-bold">Source on GitHub</a></h5>
  </footer>
);

export default Footer;
