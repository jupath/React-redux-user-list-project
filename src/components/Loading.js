import React from 'react';
import loader from '../images/spin.svg';

const Loading = () => (
  <div className="loading">
    <img src={loader} />
  </div>
);

export default Loading;