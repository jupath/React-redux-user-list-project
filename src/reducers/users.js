export const usersAreLoading = (state = false, action) => {
  switch (action.type) {
    case 'USERS_ARE_LOADING':
      return action.isLoading;
    default:
      return state;
  }
};

export const users = (state = [], action) => {
  switch (action.type) {
    case 'SET_USERS':
      return action.users;
    case 'UPDATE_USER':
      return state.map((user) => {
        if (user.username === action.username) {
          return {
            ...user,
            ...action.updatedData,
          };
        }
        return user;
      });
    default:
      return state;
  }
};
