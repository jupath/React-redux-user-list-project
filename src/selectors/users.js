
export default (users, { text, orderBy }) => {
  return users.filter(user => {
    const regex = new RegExp( text, 'gi' );
    return user.firstName.match( regex ) || user.lastName.match( regex );
  }).sort((a, b) => {
    if( orderBy === 'desc' ) {
      return a.lastName < b.lastName ? 1 : -1;
    } else if ( orderBy === 'asc' ) {
      return a.lastName < b.lastName ? -1 : 1;
    }
  });
}