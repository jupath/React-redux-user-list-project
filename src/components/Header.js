import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Input } from 'reactstrap';
import { setOrderBy, setText } from '../actions/filters';

export class Header extends Component {
  handleOrderByClick = (orderBy) => {
    this.props.setOrderBy(orderBy);
  }

  handleSetText = (event) => {
    const text = event.target.value;
    this.props.setText(text);
  }

  render() {
    return (
      <header className="header">
        <div className="container text-center py-4">
          <h1>User list built with React and Redux</h1>
          <p>This app uses async-await to fetch users from <a href="https://randomuser.me/" target="_blank">randomuser.me</a> API</p>
          <div className="px-lg-7 pb-4">
            <Input type="text" placeholder="Search by name..." className="header__searchform" onChange={this.handleSetText} />
          </div>
          <div>
            <span>Order list by name: </span>
            <Button id="ascButton" color="success" className="mr-1" disabled={this.props.orderBy === 'asc'} onClick={() => this.handleOrderByClick('asc')}>Asc</Button>
            <Button id="descButton" color="success" className="mr-1" disabled={this.props.orderBy === 'desc'} onClick={() => this.handleOrderByClick('desc')}>Desc</Button>
            <Button id="defaultButton" color="success" disabled={this.props.orderBy === undefined} onClick={() => this.handleOrderByClick()}>Default</Button>
          </div>
        </div>
      </header>
    );
  }
}

const mapStateToProps = state => ({
  orderBy: state.filters.orderBy,
});

const mapDispatchToProps = dispatch => ({
  setOrderBy: orderBy => dispatch(setOrderBy(orderBy)),
  setText: text => dispatch(setText(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
