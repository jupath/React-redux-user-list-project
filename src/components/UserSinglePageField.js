import React, { Component } from 'react';
import { connect } from 'react-redux';
import FaPencil from 'react-icons/lib/fa/pencil';
import PropTypes from 'prop-types';
import UserSinglePageForm from './UserSinglePageForm';
import { updateUser } from '../actions/users';

export class UserSinglePageField extends Component {
  state = {
    isUpdate: false,
  }

  onUpdateButton = () => {
    this.setState({
      isUpdate: true,
    });
  }

  handleUpdateUser = (updatedData) => {
    this.props.updateUser(this.props.username, { [this.props.field]: updatedData });
    this.setState({
      isUpdate: false,
    });
  }

  render() {
    let renderField;
    if (this.state.isUpdate) {
      renderField = (<UserSinglePageForm
        data={this.props.data}
        handleUpdateUser={this.handleUpdateUser}
      />);
    } else {
      renderField = (<span className="user-single__text">
        {this.props.data}
        {this.props.isEditable &&
        <a
          role="button"
          tabIndex={0}
          className="ml-2"
          onClick={this.onUpdateButton}
          onKeyDown={this.onUpdateButton}
        >
          <FaPencil />
        </a>}
      </span>);
    }

    return (
      <div>
        { renderField }
      </div>
    );
  }
}

UserSinglePageField.propTypes = {
  updateUser: PropTypes.func.isRequired,
  username: PropTypes.string.isRequired,
  field: PropTypes.string.isRequired,
  data: PropTypes.string.isRequired,
  isEditable: PropTypes.bool.isRequired,
};

const mapDispatchToProps = dispatch => ({
  updateUser: (username, user) => dispatch(updateUser(username, user)),
});

export default connect(null, mapDispatchToProps)(UserSinglePageField);
