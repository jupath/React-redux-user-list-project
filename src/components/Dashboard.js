import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterUsers from '../selectors/users';
import Header from './Header';
import User from './User';
import Loading from './Loading';

export class Dashboard extends Component {
  renderUsers = () => {
    if (this.props.isLoading) {
      return <Loading />;
    }
    const users = this.props.users.map((user) => {
      const firstName = user.firstName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);
      const lastName = user.lastName.replace(this.props.filterText, `<span class="hl">${this.props.filterText}</span>`);

      const userData = {
        firstName,
        lastName,
        image: user.image,
        username: user.username,
        cell: user.cell,
      };

      return <User key={user.username} userData={userData} />;
    });
    return users;
  }

  render() {
    return (
      <div>
        <Header />
        <div className="container">
          <div className="row users pt-5">
            { this.renderUsers() }
          </div>
        </div>
      </div>
    );
  }
}

Dashboard.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  users: filterUsers(state.users, state.filters),
  filterText: state.filters.text,
  isLoading: state.usersAreLoading,
});

export default connect(mapStateToProps)(Dashboard);
