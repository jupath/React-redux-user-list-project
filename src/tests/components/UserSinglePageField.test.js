import React from 'react';
import { shallow } from 'enzyme';
import { UserSinglePageField } from '../../components/UserSinglePageField';
import users from '../fixtures/users';

test('should render correctly when isEditable is false and isUpdate state is false ', () => {
  const wrapper = shallow(<UserSinglePageField data={users[0].email} isEditable={false} />);
  expect(wrapper).toMatchSnapshot();
});

test('should handle update button on click and render correctly', () => {
  const wrapper = shallow(<UserSinglePageField data={users[0].email} isEditable={true} />);
  wrapper.find('a').simulate('click');
  expect(wrapper.state('isUpdate')).toBe(true);
  expect(wrapper).toMatchSnapshot();
})

