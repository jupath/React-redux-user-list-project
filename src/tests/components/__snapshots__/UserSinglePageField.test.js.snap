// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle update button on click and render correctly 1`] = `
<div>
  <UserSinglePageForm
    data="romain.hoogmoed@example.com"
    handleUpdateUser={[Function]}
  />
</div>
`;

exports[`should render correctly when isEditable is false and isUpdate state is false  1`] = `
<div>
  <span
    className="user-single__text"
  >
    romain.hoogmoed@example.com
  </span>
</div>
`;
